# Dataquest project 3: Exploring Car Sales Data

We're here to explore car sales data with our newly-learned friends, NumPy and Pandas. In the dataset we're about to import, there are 50,000 datapoints across 20 columns, stuffed with errors to fix before analysis.

First, let's get some libraries.

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv("autos.csv", encoding = "Latin-1")

In [2]:
autos.head(2)

dateCrawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offerType   price   abtest vehicleType  yearOfRegistration    gearbox  \
0   Angebot  $5,000  control         bus                2004    manuell   
1   Angebot  $8,500  control   limousine                1997  automatik   

   powerPS   model   odometer  monthOfRegistration fuelType    brand  \
0      158  andere  150,000km                    3      lpg  peugeot   
1      286     7er  150,000km                    6   benzin      bmw   

  notRepairedDamage          dateCreated  nrOfPictures  postalCode  \
0              nein  2016-03-26 00:00:00             0       79588   
1              nein  2016-04-04 00:00:00             0       71034   

              lastSeen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

In [3]:
print(autos.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

### Notes

- The dataset consists of 20 columns
- 15 are strings. 5 are int, being: yearofregistration, powerPS, monthOfRegistration, nrOfPictures and lastSeen
- 5 columns have null values, all under 20% (notRepairedDamage has the most)
- camelCase is everywhere which means separating/creating new columns name requires a new appoach

## 2. Cleaning the columns
Let's clean up the column names

In [4]:
#Before
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
#Renaming
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
'odometer', 'registration_month', 'fuel_type', 'brand',
'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
'last_seen']

In [6]:
#After
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

We've simply renamed the columns here. There seemed to be more fuss than before. 2. in data cleaning basics was teaching me the strip() function, and the later rename the column comes if we've done something horrible to the data below it (such as strip out kg or lbs).

## 3. Cleaning and some analysis

Now we're looking for:
- Text columns where all/most values are the same
- Examples of numeric data stored as text which can be cleaned and converted.

#### Useful methods

- `df.describe()` (with include='all' to get both categorical and numeric columns)
- `Series.value_counts()`
- `Series.head()`

In [7]:
#Use DataFrame.describe() to look at descriptive statistics for all columns.
autos.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

### Notes:

- seller, offer_type and nr_of_pictures are all single value columns. Ab_test, gearbox, unrepaired_damage only have two values. 
- Interesting data would be likely name, price, vehicle_type, registration_year, brand and odometer (which is mileage in English, but in KM).
- Of those, price is stored as text.

I have nothing more to say on this issue. Perhaps others do more investigation, but I don't find this dataset particularly inspiring.

But I have further instruction. For price and odometer columns I have to:

- Remove any non-numeric characters.
- Convert the column to a numeric dtype.
- Use DataFrame.rename() to rename the column to odometer_km.

In [8]:
#Let's get to work on price first.

unique_price = autos['price'].unique()
print(unique_price)

['$5,000' '$8,500' '$8,990' ... '$385' '$22,200' '$16,995']


In [9]:
#There are two unique characters, $ and ,

autos["price"] = autos["price"].str.replace("$", "").str.replace(",","").astype(int)

In [10]:
#Now that we have price, let's cover odometer

unique_odometer = autos['odometer'].unique()
print(unique_odometer)

['150,000km' '70,000km' '50,000km' '80,000km' '10,000km' '30,000km'
 '125,000km' '90,000km' '20,000km' '60,000km' '5,000km' '100,000km'
 '40,000km']


In [11]:
#There are two unique characters, km and ,

autos["odometer"] = autos["odometer"].str.replace("km", "").str.replace(",","").astype(int)

In [12]:
#Renaming header of odometer

autos.rename({"odometer":"odometer_km"}, axis = 1, inplace=True)

In [13]:
#Finally, checking that we did it

autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type  price  ab_test vehicle_type  registration_year    gearbox  \
0    Angebot   5000  control          bus               2004    manuell   
1    Angebot   8500  control    limousine               1997  automatik   
2    Angebot   8990     test    limousine               2009    manuell   

   power_ps   model  odometer_km  registration_month fuel_type       brand  \
0       158  andere       150000                   3       lpg     peugeot   
1       286     7er       150000                   6    benzin         bmw   
2       102    golf        70000                   7    benzin  volkswagen   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   
2              nein  2016-03-26 00:00:00               0        35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

## 4. Exploring Odometer_km and Price

Let's "look for data that isn't right" in these two columns. This means looking for high and low values that could be outliers.

Here's what could be in use:

- Series.unique().shape to see how many unique values
- Series.describe() to view min/max/median/mean etc
- Series.value_counts(), with some variations:
    - chained to .head() if there are lots of values.
    - Because Series.value_counts() returns a series, we can use Series.sort_index() with ascending= True or False to view the highest and lowest values with their counts (can also chain to head() here).

- When removing outliers, we'll use df[df["col"].between(x,y)]

#### Odometer cleanup

In [14]:
#How many unique values are there?

unique_odo = autos['odometer_km'].unique().shape
print("There are", unique_odo, "values in odometer_km. Below is their distribution")

autos["odometer_km"].describe()

There are (13,) values in odometer_km. Below is their distribution


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [15]:
#And here you can see car mileage per segment
autos['odometer_km'].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

It doesn't seem like there are any outliers. It would naturally make sense that a high number of second-hand cars would exceed a high mileage (and doesn't take into account any other figure beyond 150,000. The same issue is found in 5000 -- no car is listed from 0-4999.

#### Price cleanup

In [16]:
unique_odo = autos['price'].unique().shape
print("There are", unique_odo, "values in price. Below is their distribution")

autos["price"].describe()

There are (2357,) values in price. Below is their distribution


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [17]:
#How many unique values are there?

autos['price'].value_counts().sort_index(ascending=False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

In [18]:
autos['price'].value_counts().sort_index(ascending=False).tail(15)

18       1
17       3
15       2
14       1
13       2
12       3
11       2
10       7
9        1
8        1
5        2
3        1
2        3
1      156
0     1421
Name: price, dtype: int64

Unlike odometer, there is a super wide distribution in price. I would remove all prices between 1-100, as they seem a little odd, but won't unless I'm told to. I will remove the higher outliers though, from 350,000 up -- as the cars in the millions will probably be joke prices.

In [19]:
#Removing joke prices with Boolean indexing

autos = autos.drop(autos[(autos["price"].between(350001,100000000))].index)

In [20]:
autos['price'].value_counts().sort_index(ascending=False).head(15)

350000    1
345000    1
299000    1
295000    1
265000    1
259000    1
250000    1
220000    1
198000    1
197000    1
194000    1
190000    1
180000    1
175000    1
169999    1
Name: price, dtype: int64

Looks like that takes care of that.

## 5. Exploring data columns

Time to look at columns and date ranges. There are 5 columns that _could_ represent dates. These are either crawled for or existed as readable on the website.

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, `date_crawled`, `last_seen`, and `ad_created` are all identified as string values, so we need to convert the data into a numerical representation.

The other two, `registration_month` and `registration year` are represented as numeric values, so we can use methods like Series.describe() to understand the distribution without extra data processing.

Let's look at their setup.

In [21]:
autos[['date_crawled','ad_created','last_seen']][0:3]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37

To understand the date range, we can:
1. Extract the date values
2. Use `Series.value_counts()` to generate a distribution
3. Sort by index

To select the first 10 characters in each column, we can use `Series.str[:10]`:

`print(autos['date_crawled'].str[:10])`

In [22]:
#Extract the date values
autos["date_crawled_day"] = (autos["date_crawled"]
                                       .str.split()
                                       .str[0])

autos["ad_created_day"] = (autos["ad_created"]
                                       .str.split()
                                       .str[0])

autos["last_seen_day"] = (autos["last_seen"]
                                       .str.split()
                                       .str[0])

In [23]:
autos["date_crawled_day"].value_counts(normalize=True, dropna=False).sort_index(ascending=True).round(2)


2016-03-05    0.03
2016-03-06    0.01
2016-03-07    0.04
2016-03-08    0.03
2016-03-09    0.03
2016-03-10    0.03
2016-03-11    0.03
2016-03-12    0.04
2016-03-13    0.02
2016-03-14    0.04
2016-03-15    0.03
2016-03-16    0.03
2016-03-17    0.03
2016-03-18    0.01
2016-03-19    0.03
2016-03-20    0.04
2016-03-21    0.04
2016-03-22    0.03
2016-03-23    0.03
2016-03-24    0.03
2016-03-25    0.03
2016-03-26    0.03
2016-03-27    0.03
2016-03-28    0.03
2016-03-29    0.03
2016-03-30    0.03
2016-03-31    0.03
2016-04-01    0.03
2016-04-02    0.04
2016-04-03    0.04
2016-04-04    0.04
2016-04-05    0.01
2016-04-06    0.00
2016-04-07    0.00
Name: date_crawled_day, dtype: float64

Not interesting data on date crawled, max 4% of data was crawled on a given day.

In [24]:
autos["ad_created_day"].value_counts(normalize=True, dropna=False).sort_index(axis=0, ascending=True).round(2)

2015-06-11    0.00
2015-08-10    0.00
2015-09-09    0.00
2015-11-10    0.00
2015-12-05    0.00
2015-12-30    0.00
2016-01-03    0.00
2016-01-07    0.00
2016-01-10    0.00
2016-01-13    0.00
2016-01-14    0.00
2016-01-16    0.00
2016-01-22    0.00
2016-01-27    0.00
2016-01-29    0.00
2016-02-01    0.00
2016-02-02    0.00
2016-02-05    0.00
2016-02-07    0.00
2016-02-08    0.00
2016-02-09    0.00
2016-02-11    0.00
2016-02-12    0.00
2016-02-14    0.00
2016-02-16    0.00
2016-02-17    0.00
2016-02-18    0.00
2016-02-19    0.00
2016-02-20    0.00
2016-02-21    0.00
              ... 
2016-03-09    0.03
2016-03-10    0.03
2016-03-11    0.03
2016-03-12    0.04
2016-03-13    0.02
2016-03-14    0.04
2016-03-15    0.03
2016-03-16    0.03
2016-03-17    0.03
2016-03-18    0.01
2016-03-19    0.03
2016-03-20    0.04
2016-03-21    0.04
2016-03-22    0.03
2016-03-23    0.03
2016-03-24    0.03
2016-03-25    0.03
2016-03-26    0.03
2016-03-27    0.03
2016-03-28    0.03
2016-03-29    0.03
2016-03-30  

In [25]:
autos["last_seen_day"].value_counts(normalize=True, dropna=False).sort_index(axis=0, ascending=True).round(2)


2016-03-05    0.00
2016-03-06    0.00
2016-03-07    0.01
2016-03-08    0.01
2016-03-09    0.01
2016-03-10    0.01
2016-03-11    0.01
2016-03-12    0.02
2016-03-13    0.01
2016-03-14    0.01
2016-03-15    0.02
2016-03-16    0.02
2016-03-17    0.03
2016-03-18    0.01
2016-03-19    0.02
2016-03-20    0.02
2016-03-21    0.02
2016-03-22    0.02
2016-03-23    0.02
2016-03-24    0.02
2016-03-25    0.02
2016-03-26    0.02
2016-03-27    0.02
2016-03-28    0.02
2016-03-29    0.02
2016-03-30    0.02
2016-03-31    0.02
2016-04-01    0.02
2016-04-02    0.02
2016-04-03    0.03
2016-04-04    0.02
2016-04-05    0.12
2016-04-06    0.22
2016-04-07    0.13
Name: last_seen_day, dtype: float64

The one above this was also too long to figure out for now. This last one though, 47% of last seen we rein the last three days of the cohort, so I guess someone bought who made the dataset bought them all up.

In [26]:
autos["registration_year"].describe()

count    49986.000000
mean      2005.075721
std        105.727161
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

There's a few outliers here, with 1000 as min and 9999 as max. That said, the middle of the range is sensible enough. We probably have to clean out those extremes too. For top, anything over 2018 would be insensible. Let's see the first dates below.

### 6. Dealing with Incorrect Registration Year Data

In [27]:
autos['registration_year'].value_counts().sort_index(ascending=False).tail(15)

1943    1
1941    2
1939    1
1938    1
1937    4
1934    2
1931    1
1929    1
1927    1
1910    9
1800    2
1500    1
1111    1
1001    1
1000    1
Name: registration_year, dtype: int64

According to wikipedia, the first car was invented in 1885, anything below 1910 is dubious (even then, why would you list you 1910 auto on eBay?) I'm going to go with deleting this. Also going beyond, 2016 is the last year of data, so nothing could be sold beyond that date

In [28]:
autos = autos[autos.registration_year.between(1900,2016)]

In [29]:
autos['registration_year'].value_counts().sort_index(ascending=False)

2016    1316
2015     399
2014     665
2013     806
2012    1323
2011    1634
2010    1597
2009    2097
2008    2231
2007    2304
2006    2707
2005    3015
2004    2737
2003    2727
2002    2533
2001    2702
2000    3354
1999    2998
1998    2453
1997    2028
1996    1444
1995    1312
1994     660
1993     445
1992     390
1991     356
1990     395
1989     181
1988     142
1987      75
        ... 
1968      26
1967      27
1966      22
1965      17
1964      12
1963       9
1962       4
1961       6
1960      33
1959       7
1958       4
1957       2
1956       5
1955       2
1954       2
1953       1
1952       1
1951       2
1950       3
1948       1
1943       1
1941       2
1939       1
1938       1
1937       4
1934       2
1931       1
1929       1
1927       1
1910       9
Name: registration_year, Length: 78, dtype: int64

In [30]:
autos["registration_year"].value_counts(ascending=True, normalize=True).round(2)

1952    0.00
1948    0.00
1929    0.00
1927    0.00
1939    0.00
1938    0.00
1943    0.00
1953    0.00
1931    0.00
1955    0.00
1951    0.00
1957    0.00
1934    0.00
1941    0.00
1954    0.00
1950    0.00
1962    0.00
1937    0.00
1958    0.00
1956    0.00
1961    0.00
1959    0.00
1963    0.00
1910    0.00
1964    0.00
1965    0.00
1969    0.00
1975    0.00
1977    0.00
1966    0.00
        ... 
1985    0.00
1988    0.00
1989    0.00
1991    0.01
1992    0.01
1990    0.01
2015    0.01
1993    0.01
1994    0.01
2014    0.01
2013    0.02
1995    0.03
2016    0.03
2012    0.03
1996    0.03
2010    0.03
2011    0.03
1997    0.04
2009    0.04
2008    0.05
2007    0.05
1998    0.05
2002    0.05
2001    0.06
2006    0.06
2003    0.06
2004    0.06
1999    0.06
2005    0.06
2000    0.07
Name: registration_year, Length: 78, dtype: float64

### 7. Exploring Price by Brand

And now we're back to loops because we're thinking about aggregation. It's natural to explore variations _across_ brands, so we go to the brand column. First let's see what brands we have.

In [31]:
autos["brand"].value_counts(ascending=False, normalize=True).round(2).head(10)

volkswagen       0.21
bmw              0.11
opel             0.11
mercedes_benz    0.10
audi             0.09
ford             0.07
renault          0.05
peugeot          0.03
fiat             0.03
seat             0.02
Name: brand, dtype: float64

These are the the top brands by percentage. VW is top at 20%, the tenth most popular is Seat with 2%. I'll probably just look at those with 5% market share and up -- would probably be interesting to compare German cars in the German market so I'm going to take top 5.

In [32]:
brand_counts = autos["brand"].value_counts(normalize=True)
top_brands = brand_counts[brand_counts > .08].index
print(top_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi'], dtype='object')


In [33]:
mean_brand_price = {}

for brand in top_brands:
    selected_rows = autos[autos["brand"] == brand]
    mean_price = selected_rows["price"].mean()
    mean_brand_price[brand] = int(mean_price)
    
print(mean_brand_price)

{'bmw': 8102, 'mercedes_benz': 8485, 'volkswagen': 5231, 'audi': 9093, 'opel': 2876}


I cheated on that last bit because I am still pretty poor at loops. Anyway, Some cars are more expensive than others judging by mean price. For our final trick, I'm going to loop over mileage data then do something new, create a dataframe with those two things together.

In [35]:
mean_mileage = {}
for brand in top_brands:
    brand_select = autos[autos["brand"] == brand]
    mean_miles = brand_select["odometer_km"].mean()
    mean_mileage[brand] = int(mean_miles)
print(mean_mileage)

{'bmw': 132431, 'mercedes_benz': 130856, 'volkswagen': 128724, 'audi': 129287, 'opel': 129223}


Now let's combine these two to compare:

In [58]:
price_series = pd.Series(mean_brand_price).sort_values(ascending=False)
mileage_series = pd.Series(mean_mileage).sort_values(ascending=False)
df = pd.DataFrame(mileage_series, columns=["mean_mileage"])
df["mean_price"] = price_series
print(df)

               mean_mileage  mean_price
bmw                  132431        8102
mercedes_benz        130856        8485
audi                 129287        9093
opel                 129223        2876
volkswagen           128724        5231


Again, I cheated here but it seems to work where other methods didn't. The analysis itself is stupid, better brands and better cars cost more. Mileage has nothing to do with it.